src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2083,6 +2083,53 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
   return this;
 }
 
+Node* MulVLNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  if (Matcher::supports_doubleword_mult_with_quadword_staturation() &&
+      !is_mult_lower_double_word()) {
+    auto is_clear_upper_doubleword_uright_shift_pattern = [](const Node* n) {
+      return n->Opcode() == Op_URShiftVL &&
+             n->in(2)->Opcode() == Op_RShiftCntV && n->in(2)->in(1)->is_Con() &&
+             n->in(2)->in(1)->bottom_type()->isa_int() &&
+             n->in(2)->in(1)->bottom_type()->is_int()->get_con() == 32;
+    };
+
+    auto is_lower_doubleword_mask_pattern = [](const Node* n) {
+      if (n->Opcode() == Op_AndV) {
+        Node* replicate_operand = n->in(1)->Opcode() == Op_Replicate ? n->in(1)
+                                  : n->in(2)->Opcode() == Op_Replicate
+                                      ? n->in(2)
+                                      : nullptr;
+        if (replicate_operand) {
+          return replicate_operand->in(1)->is_Con() &&
+                 replicate_operand->in(1)->bottom_type()->isa_long() &&
+                 replicate_operand->in(1)
+                         ->bottom_type()
+                         ->is_long()
+                         ->get_con() == 4294967295L;
+        } else {
+          return false; // Replication match failed
+        }
+      } else {
+        return false; // AndV match failed
+      }
+    };
+
+    // Detect following IR pattern for doubleword multiplication with quadword
+    // satuation.
+    // MulL ( And  SRC1,  0xFFFFFFFF)   ( And  SRC2,  0xFFFFFFFF)
+    // MulL (URShift SRC1 , 32) (URShift SRC2, 32)
+    // MulL (URShift SRC1 , 32)  ( And  SRC2,  0xFFFFFFFF)
+    // MulL ( And  SRC1,  0xFFFFFFFF) (URShift SRC2 , 32)
```

Comment:
I don't understand how it works... According to the `VPMULDQ`/`VPMULUDQ` consumes a vector of double words and produces a vector of quadwords. But it looks like `SRC1`/`SRC2` are always vectors of longs (quadwords). And `vmuludq_reg` in `x86.ad` just takes the immedate operands and pass them into `vpmuludq` which doesn't look right...

Reply:
`vpmuludq` does a long multiplication but throws away the upper bits of the operands, effectively does a `(x & max_juint) * (y & max_juint)`

Reply:
You can see its pseudocode here https://www.felixcloutier.com/x86/pmuludq

    VPMULUDQ (VEX.256 Encoded Version)[ ¶](https://www.felixcloutier.com/x86/pmuludq#vpmuludq--vex-256-encoded-version-)
    DEST[63:0] := SRC1[31:0] * SRC2[31:0]
    DEST[127:64] := SRC1[95:64] * SRC2[95:64]
    DEST[191:128] := SRC1[159:128] * SRC2[159:128]
    DEST[255:192] := SRC1[223:192] * SRC2[223:192]
    DEST[MAXVL-1:256] := 0

Reply:
Got it. Now it makes perfect sense. Thanks for the clarifications!

Reply:
Actually, it makes detecting the pattern during matching even simpler than I initially thought. Since there's no need to match any non-trivial ideal IR tree, AD instruction can just match a single `MulVL`, but detect operand shapes using a predicate.